devlib/module/cgroups2.py

```diff
@@ -0,0 +1,1972 @@
+#    Copyright 2022 ARM Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""
+
+Successor to the ``cgroups`` devlib module.
+
+This one handles both cgroups V1 and V2 transparently with an API matching
+cgroup v2 semantic.
+
+It also handles the cgroup delegation API of systemd.
+
+
+.. code-block:: python
+
+    # Necessary Imports
+    from devlib import Target
+    from devlib.module.cgroups2 import RequestTree
+
+    # Connecting to target device. Configure appropriately.
+
+    my_target = Target(name="target",
+                        host="127.0.0.1",
+                        port="0000",
+                        username="root",
+                        password="root",
+                        kind="linux")
+
+    # Instantiating the RequestTree object,
+    # representing a hierarchical CGroup structure consisting
+    # of a singular parent and child CGroup relationship.
+
+    request = RequestTree(
+        name="root",
+        children=[
+            RequestTree(
+                name="child",
+                controllers={"pids": {"max": 10}}
+            )
+        ],
+        controllers={"pids": {"max": 20}},
+    )
+
+    # Printing the request to display/inspect the hierarchical
+    # tree-like structure of the RequestTree object.
+
+    print(request)
+
+    '''
+    └──root (pids) {'max': 20}
+        └──child (pids) {'max': 10}
+    '''
+
+    # To set-up either CGroup version hierarchies, ensure the target device is
+    # appropriately configured alongside a CGroup version appropriate RequestTree object.
+
+    # Setting up the RequestTree object CGroup hierarchy onto target device
+    # as a V1 hierarchy, and printing the returned ResponseTree object.
+
+    with request.setup_hierarchy(target=my_target, version=1) as CGroup_hierarchy:
```

Comment:
Hi Marco,

Yes it is possible to auto-detect which version a controller is mounted on as and dealing with this automatically of-course, however there is a big issue which arises in the case when mounting a user requested CGroup hierarchy on a Linux-Target with no CGroups pre-mounted. 
In this case, the version that a CGroup Controller should be mounted as would be undefined. Of-course we could decide that either a V1 or V2 CGroup hierarchy is automatically hard-coded to be mounted in this case, however this would bring in a level of obscurity when using the API from the standpoint of the user given that certain controllers and controller interface file aren't identical between the 2 CGroup versions. This was a conscious design choice in order to combat this and to give the responsibility to the user.

I just want to clarify a point when it comes to handling both versions transparently.  The aim of the module is to allow the User of the API to define, mount and interact with a  CGroup hierarchy using the semantics and rules introduced with CGroups V2, as seen with the `RequestTree` and `ResponseTree` classes respectively.

Reply:
Ah I see, thanks for the explanation I think that seems like a reasonable course of action.